building-and-packaging/material/cmake.md

```diff
@@ -0,0 +1,414 @@
+...
++ sse();
++ std::cout << precice::getVersionInformation() << std::endl;
+```
+
+`CMakeLists.txt`:
+
+```diff
+find_package(precice REQUIRED CONFIG)
+target_link_libraries("${PROJECT_NAME}" PRIVATE precice::precice)
+```
+
+- `find_package` works if dependency is "cmake-ready"
+- Alternative option if lib is in a discoverable location (standard system path or adjusted `LD_LIBRARY_PATH`)
+
+```cmake
+find_library(precice REQUIRED CONFIG)
+target_link_libraries("${PROJECT_NAME}" PRIVATE precice)
+```
+
+- TODO: external dependencies via pkg-config?
+    - @fsimonis: Is `find_package` already using pkg-config? Or is there a real third option?
+    - include(..), pkg-config.search_modules ... link ... cmake searches in pkg-config
+
+### Options and variables
+
+- For good software, a user should not need to touch `CMakeLists.txt`
+- Use variables to talk to cmake
+- Let's try to make preCICE an optional dependency of our code (not really everybody has preCICE installed, sadly)
+
+`CMakeLists.txt`:
+
+```cmake
+if (ENABLE_PRECICE)
+  message(STATUS "preCICE enabled")
+  find_library(precice REQUIRED CONFIG)
+  target_link_libraries("${PROJECT_NAME}" PRIVATE precice)
+else()
+  message(STATUS "preCICE disabled")
+  target_compile_definitions("HelloWorld" PRIVATE NO_PRECICE)
+endif()
+```
+
+`main.cpp`:
+
+```c++
+#ifndef NO_PRECICE
+  std::cout << precice::getVersionInformation() << std::endl;
+#endif
+```
+
+- Build without preCICE `cmake -DENABLE_PRECICE=OFF ..`
+- Create different folder and build there with preCICE
+
+TODO: variables, which ones are import? how to pass? change compiler etc
+
+`CXX ?= g++`
+
+TODO: More reading on this:
+
+- https://cmake.org/cmake/help/latest/module/FindPkgConfig.html#command:pkg_check_modules
+- https://cmake.org/cmake/help/latest/manual/cmake-variables.7.html
+- https://cmake.org/cmake/help/latest/manual/cmake-properties.7.html
+
+
+### ccmake
+
+`sudo apt-get install cmake-curses-gui`
+
+- `ccmake ..`
+- `c`, look add description of ENABLE_PRECICE`, disable it
+- `t`, show all options, helpful to find out which variables exist
+- `g`, generate `Makefile`
+
+TODO: watch some tutorial, learn more
+
+## Explain preCICE `CMakeLists.txt`
+
+> 10 mins
+
+- `sources.cmake` to include sources
+
+## Exercise
+
+TODO
+
+> 90 mins
+
+- Skeleton: main file, 2-3 other source files, each one with one dependency (e.g. Eigen (a bit difficult), deal.II, boost, mpi (?),..., boost-dev (header only) -> `boost_container_flat_set` example?)
+- Other header are initially commented out in main file.
+- Write docker file and build container, based on Ubuntu 20:04 (we only give high-level instructions this time)
+- Students write basic `CMakeLists.txt`.
+- Add dependencies step by step. (in code, in `CMakeLists.txt`, in docker recipe (we provide package names)
+- Make dependencies all optional
+- Use CTest (we already provide test stubs)
+- Open PR with docker recipe, `CMakeLists.txt, and modified code
```

Comment:
```suggestion
- Open PR with docker recipe, `CMakeLists.txt`, and modified code
```

How many Docker recipes are the students supposed to create? Maybe they should set up their repository with the `CMakeLists.txt` while maintaining notes for the commands needed to compile their package first. Afterwards, they can create a Dockerfile based on these notes. 

What should the Dockerfile be good for in the end? Is it just about repetition?

Reply:
Two-fold: repetition and reproducibility (also for us when checking)